src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,421 @@
+     * manually you can set gcProbablity to 0.0 and optionally leverage
+     * self::createTtlIndex(int $expireAfterSeconds = 0).
+     *
+     * writeConcern, readConcern and readPreference are not specified by
+     * MongoDbStore meaning the collection's settings will take effect.
+     * @see https://docs.mongodb.com/manual/applications/replication/
+     */
+    public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)
+    {
+        $this->mongo = $mongo;
+
+        $this->initialTtl = $initialTtl;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+            'gcProbablity' => 0.001,
+        ), $options);
+
+        if (!isset($this->options['database'])) {
+            throw new InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        if ($this->options['gcProbablity'] < 0.0 || $this->options['gcProbablity'] > 1.0) {
+            throw new InvalidArgumentException(sprintf(
+                '"%s" gcProbablity must be a float from 0.0 to 1.0, "%f" given.',
+                __METHOD__,
+                $this->options['gcProbablity']
+            ));
+        }
+
+        if ($this->initialTtl <= 0) {
+            throw new InvalidArgumentException(sprintf(
+                '%s() expects a strictly positive TTL. Got %d.',
+                __METHOD__,
+                $this->initialTtl
+            ));
+        }
+    }
+
+    /**
+     * Create a TTL index to automatically remove expired locks.
+     *
+     * If the gcProbablity option is set higher than 0.0 (defaults to 0.001);
+     * there is a chance this will be called on self::save().
+     *
+     * Otherwise; this should be called once manually during database setup.
+     *
+     * Alternatively the TTL index can be created manually on the database:
+     *
+     *  db.lock.ensureIndex(
+     *      { "expires_at": 1 },
+     *      { "expireAfterSeconds": 0 }
+     *  )
+     *
+     * Please note, expires_at is based on the application server. If the
+     * database time differs; a lock could be cleaned up before it has expired.
+     * To ensure locks don't expire prematurely; the lock TTL should be set
+     * with enough extra time to account for any clock drift between nodes.
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     *
+     * @return string The name of the created index
+     *
+     * @throws UnsupportedException          if options are not supported by the selected server
+     * @throws MongoInvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException        for other driver errors (e.g. connection errors)
+     */
+    public function createTtlIndex(int $expireAfterSeconds = 0): string
```

Comment:
This method should be either 
- public: must be called by user (and disabled the GC logic when version >= 2.2)
- private: automatically called by GC

Reply:
My intent was to make default behaviour to create a TTL Index with a `gcProbablity` default of `0.001`. This is to prevent orphan locks for anyone who uses this store without reading docs or changing any config.

If someone doesn't want this to happen (and they read the docs), I left the `createTtlIndex` method `public` allowing someone to configure `gcProbablity = 0.0` and hook `createTtlIndex` in their deployment or database set up scripts.

See constructor docs:

     * .... If you prefer to create your TTL Index
     * manually you can set gcProbablity to 0.0 and optionally leverage
     * self::createTtlIndex(int $expireAfterSeconds = 0).

It's my opinion setting TTL index automatically is dangerous (if someone has differing clocks between DBMS and App) as well as an unnecessary expense. I would like to make a recommendation in https://github.com/symfony/symfony-docs/pull/9807 to set `gcProbablity = 0.0` and manually create the TTLIndex.

If you disagree this this approach, which way do you think is the better way to go?

Reply:
:+1: I forgot the part about GC=0 and creting the index manually. Sound's good to me